blog: goose janitor (#4131)
block/goose: documentation/blog/2025-08-18-ai-teammate/index.md

<!-- truncate -->

## Phase 1: Making Knowledge AI-Accessible

Gemma's first move was brilliant in its simplicity. She documented how to safely remove feature flags from the repository. She also structured the documentation so AI agents could follow it too!

This wasn't just another README file. This was documentation designed for both human and AI to understand and act upon.

## Phase 2: Validation Across Generations

She didn't stop there. Gemma put her documentation to the test, using Cursor AI to validate the approach across three different generations of legacy feature flag implementations:

- super legacy flags the team used way back yonder
- legacy flags that were newer but are now old
- current implementation of feature flags

By testing across this spectrum of legacy code, she proved that the documentation worked. AI could actually understand and safely navigate the complexity of real world legacy systems.

## Phase 3: Building an AI Team Member

This was great progress. The documentation alone would help teams clean up faster. Gemma could have stopped here.

But instead, she built something remarkable: **Goose Janitor**.

> Goose Janitor acts as new AI team member whose responsibility is to tidy up the code after we're done experimenting. Gemma modified an existing [Goose recipe](/recipes/detail/?id=clean-up-feature-flag) that autonomously removes feature flags from the codebase. Here's how it works:

Reviewer: ```suggestion
Goose Janitor acts as new AI team member whose responsibility is to tidy up the code after we're done experimenting. Gemma drew inspiration from existing [Goose recipes](/recipes/detail/?id=clean-up-feature-flag) and internal talks from other departments to create it. Here's how it works:
```

> Goose Janitor isn't just a tool, it's a new AI team member whose responsibility is to tidy up the code after we're done experimenting. Gemma modified an existing [Goose recipe](/recipes/detail/?id=clean-up-feature-flag) that autonomously removes feature flags from the codebase. Here's how it works:

Reviewer 2: LOL a nitpick 
```suggestion
Goose Janitor acts as new AI team member whose responsibility is to tidy up the code after we're done experimenting. Gemma modified an existing [Goose recipe](/recipes/detail/?id=clean-up-feature-flag) that autonomously removes feature flags from the codebase. Here's how it works:
```

```bash
goose run \
--recipe .goose/recipes/goose-janitor-flag-removal.yaml \
--params feature_flag_key=log-observer-is-enabled \
--params variant_to_remain=true \
--params create_pr=false
```

> The AI agent:

Reviewer: ```suggestion
The recipe:
```

- Runs completely autonomously (no human intervention needed)
- Handles simple and complex flags 
- Uses git worktrees for safe parallel work
- Can automatically create draft pull requests

> - Integrates with a custom Xcode MCP server for deep iOS understanding

Reviewer: ```suggestion
- Integrates with [Xcode Index MCP](https://github.com/block/xcode-index-mcp) for deep iOS project understanding
```
